prov/verbs/src/verbs_profile.c

```diff
@@ -0,0 +1,625 @@
+
+static void
+vrb_prof_report_func_time(char *buf, size_t len)
+{
+	int i;
+	struct vrb_prof_time_entry *entry;
+
+	if (!vrb_prof_num_funcs)
+		return;
+	ofi_strncatf(buf, len, "Function\n");	
+	ofi_strncatf(buf, len, VRB_PROF_STR_FORMAT4,
+		"Name", "min(ns)", "max(ns)", "avg(ns)", "count");
+	for (i = 0; i < vrb_prof_num_funcs; i++) {
+		entry = &(vrb_prof_func_table[i]);
+		if (!entry->count) 
+			continue;
+		ofi_strncatf(buf, len, VRB_PROF_DATA_FORMAT4,
+			     entry->name, entry->min_tm, entry->max_tm, 
+			     (entry->total/entry->count), entry->count);
+	}
+}
+
+int vrb_prof_create(struct vrb_profile **prof)
+{
+	*prof =  calloc(1, sizeof(vrb_profile_t));
+	if (!*prof)
+		return -ENOMEM;
+
+	(*prof)->vars = vrb_prof_get_vars();
+	if (!(*prof)->vars) {
+		free(*prof);
+		*prof = NULL;
+		return -ENOMEM;
+	}
+	return 0;
+}
+
+void vrb_prof_init_state(struct vrb_profile *prof, uint64_t cur_time, int type)
+{
+	prof->state = vrb_prof_get_state_time(type);
+	if (prof->state)
+		prof->state->start = cur_time;
+}
+
+static inline bool
+vrb_prof_find_func(const char *fname, bool is_add, int *idx)
+{
+	int i;
+	
+	ofi_mutex_lock(&vrb_prof_func_mutex);
+	for (i = vrb_prof_num_funcs - 1; i >= 0; i--) {
+		if (!strncmp(vrb_prof_func_table[i].name, fname, strlen(fname))) {
+			*idx = i;
+			ofi_mutex_unlock(&vrb_prof_func_mutex);
+			return true;
+		}
+	}
+	if (is_add) {
+		memset(&(vrb_prof_func_table[vrb_prof_num_funcs]), 0,
+		       sizeof(struct vrb_prof_time_entry));
+		strncpy(vrb_prof_func_table[vrb_prof_num_funcs].name, 
+			fname, VRB_FNAME_LEN);
+		vrb_prof_func_table[vrb_prof_num_funcs].min_tm = UINT64_MAX;
+		*idx = vrb_prof_num_funcs++;
+	}
+	ofi_mutex_unlock(&vrb_prof_func_mutex);
+	return false;
+}
+
+void vrb_prof_add_func(const char *fname)
+{
+	int idx;
+	struct vrb_prof_time_entry *entry;
+
+	vrb_prof_find_func(fname, 1, &idx);
+	entry = &(vrb_prof_func_table[idx]);
+
+	entry->start = ofi_gettime_ns();
+}
+
+void vrb_prof_set_func_time(const char *fname)
+{
+	uint64_t cur_time = ofi_gettime_ns();
+	int idx;
+	bool found = vrb_prof_find_func(fname, 0, &idx);
+	struct vrb_prof_time_entry *entry;
+	uint64_t dur;
+
+	if (!found || !(vrb_prof_func_table[idx].start))
+		return;
+
+	entry = &(vrb_prof_func_table[idx]);
+	dur = cur_time - entry->start;
+	entry->min_tm = (dur < entry->min_tm) ? dur : entry->min_tm;
+	entry->max_tm = (dur > entry->max_tm) ? dur : entry->max_tm;
+	entry->total += dur;
+	entry->count++;
+}
+
+// This is called in the verbs provider VERBS_INI time.
+void vrb_prof_init()
+{
+	int size = (sizeof(vrb_prof_state_time_table) /
+		   sizeof(struct vrb_prof_state_entry));
+	int i;
+
+	ofi_prof_sys_init();
+
+	for (i = 0; i < size; i++)
+		ofi_mutex_init(&(vrb_prof_state_time_table[i].mutex));
+
+	ofi_mutex_init(&vrb_prof_counter_table.mutex);
+	ofi_mutex_init(&vrb_prof_func_mutex);
+}
+
+void vrb_prof_report(struct vrb_profile *prof)
+{
+
+	int bufsize = 4096;
+	char buf[4096];
```

Comment:
maybe allocate buf dynamically? that would allow adjusting the size at runtime in the future.